### DISCUSSION TITLE
[FR] client-to-client defaults

### DISCUSSION THREAD

**Discussion by devZer0** [pivpn/pivpn#1399] [Feature requests]
apparently,  pipvpn comes with "client-to-client" preconfigured, so by default any connected client may reach any other client - without packets passing the linux kernel tcp/ip stack.

anyhow, i tried disabling this to prevent inter-client communication, but this still works afterwards.

i think this could lead to security issues in some environments and make spreading of virus and worms easier, so imho this should 

a) better be off (secure) by default
and 
b) configurable at setup time

**Comment by orazioedoardo** [comment]
> without packets passing the linux kernel tcp/ip stack.

How? They must pass through the raspberry pi to reach other clients.

> i tried disabling this to prevent inter-client communication, but this still works afterwards.

I will check later. Probably you can enforce the block at iptables level `iptables -I FORWARD -s 10.8.0.0/24 -d 10.8.0.0/24 -j DROP`.

**Comment by devZer0** [comment]
> How? They must pass through the raspberry pi to reach other clients.

i meant this:
https://serverfault.com/questions/736274/openvpn-client-to-client
"If client-to-client is enabled, the VPN server forwards client-to-client packets internally without sending them to the IP layer of the host (i.e. to the kernel). The host networking stack does not see those packets at all."

**Comment by orazioedoardo** [comment]
Wow that’s very interesting, thanks! However, removing `client-to-client` would not increase security unless you also set up some firewall rules.

**Comment by TinCanTech** [comment]
The only `iptables` rule required is: 
`-A FORWARD -s $VPN_CIDR -d $VPN_CIDR -j DROP`

**Comment by devZer0** [comment]
yes, and i think it would be a good contermeasure if 

- no client-to-client setting
- iptables rule like above

would be the default, as this is more secure

i think it could be configurable at setup-time this way:

would you like:

1 ( ) default - secure option, no routing/connection between clients
2 ( ) all vpn clients being able to connect to each other without routing/firewall at the linux kernel level
3 ( ) same as above, but with kernel level routing in place (need to configure iptables firewalling yourself)

**Comment by orazioedoardo** [comment]
@TinCanTech Yes, I used insert (at the top) instead of append to make sure the traffic is not accepted by some rule above.

@devZer0 I think options 2 and 3 are too hard to understand considered the audience PiVPN is aimed to. 2 seems the easiest.

**Comment by devZer0** [comment]
if 2 and 3 are too hard to understand and 1 is the default - then they can simply ignore it. 

i think someone who is not willingly to read into openvpn documentation and learn about how things work should not setup vpn at all, not even with pivpn. 

for me, pivpn simplifiest openvpn setup. that does not mean you should ignore technical details or security

if you drive a car you need to learn about traffic rules and some car basics.

if you run a vpn, you should know some basics about about vpn and routing, too.

don't hide too much from the end user and give him a false impression of simplicity

**Comment by orazioedoardo** [comment]
I always thought PiVPN as a script to abstract away the complexity of a VPN installation, if the user is willing to read the openvpn documentation, probably he can go through all steps by himself.
If you go to the pivpn subreddit you realize that most users don't even read the FAQ (almost no posts with the debug and none with the troubleshooting).
Also 2 and 3 achieve the same. If the user is not confident with openvpn, but he/she wants intercommunication, the choice is confusing. How would you write "kernel level routing" in a user friendly way? 3 could be a comment to the server configuration, similar to duplicate-cn.

**Comment by TinCanTech** [comment]
> I always thought PiVPN as a script to abstract away the complexity of a VPN installation

Precisely.

At best PiVPN could offer "client to client" On|Off.
Anything more complex is out-of-scope for a script. 
(Unless the script is specifically written with complex Firewall rules in mind)

**Comment by devZer0** [comment]
mh, what about this?

1 (*) no routing/connection between clients possible (secure default option) 
2 ( ) allow communication between vpn clients, routed by openvpn software itself. *
3 ( ) allow communication between vpn clients, routed by the underlying linux os ( you can apply firewall/iptable rules here) *

* may be insecure, as some worm/virus infection may spread from client to client if there is no firewall and up2date operating system. 

otherwise i'm ok with simple option "client to client" on|off - but mind that you need to remove configuration line in server.conf AND add iptable rule to block

**Comment by orazioedoardo** [comment]
I'm ok with the ON/OFF switch:

ON: just use `client-to-client`
OFF: no `client-to-client` + iptables or ufw to drop traffic.
